File: Sources/RedlineCore/UsageStore.swift
// Parses Claude Code transcripts (~/.claude/projects/**/*.jsonl) into usage entries, with
// per-file caching keyed on mtime+size+cutoff so unchanged files are not re-read.
import Foundation

public final class UsageStore {
    public static let provider = "Claude"
    private let root: URL
    private var fileCache:
        [String: (mtime: Date, size: Int, cutoff: Date, entries: [Entry])] = [:]
    private let isoFrac: ISO8601DateFormatter = {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return f
    }()
    private let iso: ISO8601DateFormatter = {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime]
        return f
    }()

    public init(root: URL? = nil) {
        self.root = root ?? FileManager.default.homeDirectoryForCurrentUser
            .appendingPathComponent(".claude/projects")
    }

    public func scan(lookbackDays: Int, now: Date = Date()) -> [Entry] {
        let fm = FileManager.default
        let cutoff = now.addingTimeInterval(-Double(lookbackDays + 1) * 86400)
        var livePaths = Set<String>()

        guard fm.fileExists(atPath: root.path),
              let en = fm.enumerator(at: root, includingPropertiesForKeys:
                [.contentModificationDateKey, .fileSizeKey]) else { return [] }

        for case let url as URL in en {
            guard url.pathExtension == "jsonl" else { continue }
            guard let vals = try? url.resourceValues(
                forKeys: [.contentModificationDateKey, .fileSizeKey]),
                let mtime = vals.contentModificationDate,
                let size = vals.fileSize, mtime > cutoff else { continue }
            let path = url.path
            livePaths.insert(path)
            // The cutoff is part of the key because entries are filtered at parse time: a
            // cache built for 14 days answered a 30 day scan with 14 days of data.
            if let c = fileCache[path], c.mtime == mtime, c.size == size,
               c.cutoff <= cutoff { continue }
            fileCache[path] = (mtime, size, cutoff, parse(url: url, cutoff: cutoff))
        }
        fileCache = fileCache.filter { livePaths.contains($0.key) }

        // Dedup across files: resumed sessions copy identical message ids between transcripts
        var seen = Set<String>()
        var out: [Entry] = []
        for (_, c) in fileCache {
            // A cache parsed for a wider window reaches past this scan's cutoff, so the
            // window is enforced again here rather than trusted from the parse
            for e in c.entries where e.ts > cutoff {
                if let k = e.key {
                    if seen.contains(k) { continue }
                    seen.insert(k)
                }
                out.append(e)
            }
        }
        return out
    }

    func parse(url: URL, cutoff: Date) -> [Entry] {
        guard let text = try? String(contentsOf: url, encoding: .utf8) else { return [] }
        var entries: [Entry] = []
        text.enumerateLines { line, _ in
            guard line.contains("\"usage\""), let data = line.data(using: .utf8),
                  let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let msg = obj["message"] as? [String: Any],
                  let usage = msg["usage"] as? [String: Any],
                  let tsStr = obj["timestamp"] as? String,
                  let ts = self.isoFrac.date(from: tsStr) ?? self.iso.date(from: tsStr),
                  ts > cutoff else { return }
            let model = (msg["model"] as? String) ?? "unknown"
            guard model != "<synthetic>" else { return }

            let input = (usage["input_tokens"] as? Int) ?? 0
            let output = (usage["output_tokens"] as? Int) ?? 0
            let cacheRead = (usage["cache_read_input_tokens"] as? Int) ?? 0
            var c5m = (usage["cache_creation_input_tokens"] as? Int) ?? 0
            var c1h = 0
            if let cc = usage["cache_creation"] as? [String: Any] {
                c5m = (cc["ephemeral_5m_input_tokens"] as? Int) ?? 0
                c1h = (cc["ephemeral_1h_input_tokens"] as? Int) ?? 0
            }
            guard input + output + cacheRead + c5m + c1h > 0 else { return }

            var key: String? = nil
            if let mid = msg["id"] as? String {
                key = mid + ":" + ((obj["requestId"] as? String) ?? "")
            }
            entries.append(Entry(provider: UsageStore.provider, key: key, ts: ts,
                                 model: model, input: input, output: output,
                                 cacheRead: cacheRead, cache5m: c5m, cache1h: c1h))
        }
        return entries
    }
}

File: windows/src/Redline.Core/UsageStore.cs
// Parses Claude Code transcripts (~/.claude/projects/**/*.jsonl) into usage entries. Ingest tails
// into the warehouse (Keep Local History); Scan parses the whole window into a per-file cache.
using System.Text.Json.Nodes;

namespace Redline.Core;

public sealed class UsageStore
{
    public const string Provider = "Claude";
    private readonly string _root;
    private Dictionary<string, (DateTimeOffset Mtime, long Size, DateTimeOffset Cutoff, List<Entry> Entries)> _fileCache = new();

    public UsageStore(string? root = null)
    {
        _root = root ?? RedlineHome.PathFor(".claude/projects");
    }

    /// Every transcript under a root, recursively and including hidden ones, as Swift walks it.
    internal static IEnumerable<FileInfo> Transcripts(string root)
    {
        if (!Directory.Exists(root)) yield break;
        var opts = new EnumerationOptions
        {
            RecurseSubdirectories = true, IgnoreInaccessible = true, AttributesToSkip = 0,
        };
        IEnumerable<string> files;
        try { files = Directory.EnumerateFiles(root, "*", opts).ToList(); }
        catch { yield break; }
        foreach (var f in files)
        {
            if (!string.Equals(Path.GetExtension(f), ".jsonl", StringComparison.Ordinal)) continue;
            FileInfo info;
            try { info = new FileInfo(f); if (!info.Exists) continue; _ = info.Length; }
            catch { continue; }
            yield return info;
        }
    }

    /// Reads what is new in every transcript and stores it; returns records added. No lookback
    /// applies: a transcript is read once and kept until retention removes it.
    public int Ingest(Warehouse warehouse, DateTimeOffset? now = null)
    {
        var at = now ?? DateTimeOffset.UtcNow;
        if (!Directory.Exists(_root)) return 0;
        var live = new HashSet<string>();
        var added = 0;
        foreach (var info in Transcripts(_root))
        {
            var mtime = new DateTimeOffset(info.LastWriteTimeUtc, TimeSpan.Zero);
            var size = info.Length;
            var path = info.FullName;
            live.Add(path);
            var mark = warehouse.IngestMark(path);
            var start = TranscriptTail.StartOffset(mark, size);
            // Nothing appended since the last pass: this turns a huge corpus into a directory walk
            if (mark is not null && start == mark.ByteOffset && size == mark.Size) continue;

            var batch = new List<Entry>();
            var next = TranscriptTail.Read(path, start, (line, offset) =>
            {
                if (EntryFrom(line, path, offset) is { } e) batch.Add(e);
            });
            added += warehouse.Ingest(batch);
            warehouse.SetIngestMark(new IngestMark(path, Provider, size, next, mtime), at);
        }
        warehouse.ForgetIngestMarks(live, Provider);
        return added;
    }

    public List<Entry> Scan(int lookbackDays, DateTimeOffset? now = null)
    {
        var t = now ?? DateTimeOffset.UtcNow;
        var cutoff = t.AddSeconds(-(double)(lookbackDays + 1) * 86400);
        var livePaths = new HashSet<string>();
        if (!Directory.Exists(_root)) return new List<Entry>();

        foreach (var info in Transcripts(_root))
        {
            var mtime = new DateTimeOffset(info.LastWriteTimeUtc, TimeSpan.Zero);
            var size = info.Length;
            if (!(mtime > cutoff)) continue;
            var path = info.FullName;
            livePaths.Add(path);
            // The cutoff is part of the key: a cache built for 14 days must not answer a 30 day scan
            if (_fileCache.TryGetValue(path, out var c) && c.Mtime == mtime && c.Size == size &&
                c.Cutoff <= cutoff) continue;
            _fileCache[path] = (mtime, size, cutoff, Parse(path, cutoff));
        }
        _fileCache = _fileCache.Where(kv => livePaths.Contains(kv.Key)).ToDictionary(kv => kv.Key, kv => kv.Value);

        // Dedup across files: resumed sessions copy identical message ids between transcripts
        var seen = new HashSet<string>();
        var output = new List<Entry>();
        foreach (var (_, c) in _fileCache)
        {
            // A cache parsed for a wider window reaches past this cutoff, so it is enforced again
            foreach (var e in c.Entries)
            {
                if (!(e.Ts > cutoff)) continue;
                if (e.Key is { } k && !seen.Add(k)) continue;
                output.Add(e);
            }
        }
        return output;
    }

    internal List<Entry> Parse(string path, DateTimeOffset cutoff)
    {
        string text;
        try { text = File.ReadAllText(path); } catch { return new List<Entry>(); }
        var entries = new List<Entry>();
        // Counted per file, not logged per line, so a format change is one event, not thousands
        var malformed = 0;
        using var reader = new StringReader(text);
        string? line;
        while ((line = reader.ReadLine()) is not null)
        {
            if (EntryFrom(line, path, null) is { } e)
            {
                if (e.Ts > cutoff) entries.Add(e);
                continue;
            }
            // Only invalid JSON counts; a line with no usage or a synthetic model is a normal skip
            if (line.Contains("\"usage\"") && Json.Parse(line) is null) malformed++;
        }
        if (malformed > 0)
        {
            Diag.Log.Warn("transcript.lines_unparsed", "lines carried usage but were not valid JSON",
                new() { ["path"] = path, ["count"] = malformed.ToString() });
        }
        return entries;
    }

    /// One transcript line to one usage record, or null for the many lines with no usage.
    /// `offset` is where the line started and becomes the record's origin.
    internal Entry? EntryFrom(string line, string path, long? offset)
    {
        if (!line.Contains("\"usage\"")) return null;
        if (Json.ParseObject(line) is not { } obj ||
            obj["message"] is not JsonObject msg ||
            msg["usage"] is not JsonObject usage ||
            TranscriptTail.ParseTimestamp(Json.Str(obj["timestamp"])) is not { } ts) return null;
        var model = Json.Str(msg["model"]) ?? "unknown";
        if (model == "<synthetic>") return null;

        var input = Json.Int(usage["input_tokens"]) ?? 0;
        var output = Json.Int(usage["output_tokens"]) ?? 0;
        var cacheRead = Json.Int(usage["cache_read_input_tokens"]) ?? 0;
        var c5m = Json.Int(usage["cache_creation_input_tokens"]) ?? 0;
        var c1h = 0;
        if (usage["cache_creation"] is JsonObject cc)
        {
            c5m = Json.Int(cc["ephemeral_5m_input_tokens"]) ?? 0;
            c1h = Json.Int(cc["ephemeral_1h_input_tokens"]) ?? 0;
        }
        if (input + output + cacheRead + c5m + c1h <= 0) return null;

        string? key = null;
        if (Json.Str(msg["id"]) is { } mid) key = mid + ":" + (Json.Str(obj["requestId"]) ?? "");
        return new Entry(Provider, key, ts, model, input, output, cacheRead, c5m, c1h,
                         offset is { } o ? $"{path}#{o}" : null);
    }
}
